[ONNX] Add huggingface models into CI tests

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107411
* __->__ #107247

1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

titaiwangms committed:
[ONNX] Add huggingface models into CI tests

[ONNX] Add transformers models into no runtime test of fx exporter

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -422,6 +422,143 @@ def test_fake_tensor_mode_huggingface_bigscience__bloom_560m(self):
+        with torch.onnx.enable_fake_mode() as fake_context:
+            tokenizer = AutoTokenizer.from_pretrained(model_name)
+            inputs = tokenizer("Hello world!", return_tensors="pt")
+            model = AutoModel.from_pretrained(model_name)
```
For all `from_pretrained` model tests, you'd need to add them to docker 

https://github.com/pytorch/pytorch/blob/e1ee10e6f59d43cd7c0fc830954220558beb76f4/.ci/docker/common/install_onnx.sh#L39

titaiwangms replied:
Thanks! I will have another one uploading model list and add models to cache. This PR mostly logs my progress on Config refactoring.

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


[ONNX] Add transformers models into no runtime test of fx exporter

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


[ONNX] Add transformers models into no runtime test of fx exporter

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -33,10 +33,19 @@ pip_install onnx-weekly==1.15.0.dev20230717
+as_jenkins echo 'transformers.AutoModelForCausalLM.from_pretrained("mosaicml/mpt-7b", trust_remote_code=True); transformers.AutoTokenizer.from_pretrained("mosaicml/mpt-7b", trust_remote_code=True);' >> "${IMPORT_SCRIPT_FILENAME}"
+as_jenkins echo 'transformers.AutoModel.from_pretrained("openai/whisper-tiny"); transformers.WhisperConfig.from_pretrained("openai/whisper-tiny");transformers.WhisperProcessor.from_pretrained("openai/whisper-tiny");' >> "${IMPORT_SCRIPT_FILENAME}"
+as_jenkins echo 'transformers.AutoModel.from_pretrained("google/flan-t5-small"); transformers.AutoTokenizer.from_pretrained("google/flan-t5-small");' >> "${IMPORT_SCRIPT_FILENAME}"
+as_jenkins echo 'transformers.AutoModel.from_pretrained("databricks/dolly-v2-3b"); transformers.AutoTokenizer.from_pretrained("databricks/dolly-v2-3b");' >> "${IMPORT_SCRIPT_FILENAME}"
```
Maybe cache models after installing PyTorch. It might save some self-installation on dependencies.

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -33,10 +33,16 @@ pip_install onnx-weekly==1.15.0.dev20230717
+as_jenkins echo 'import transformers; transformers.AutoModel.from_pretrained("sshleifer/tiny-gpt2").to("cpu"); transformers.AutoTokenizer.from_pretrained("sshleifer/tiny-gpt2");' \
+                'transformers.AutoModel.from_pretrained("bigscience/bloom-560m").to("cpu"); transformers.AutoTokenizer.from_pretrained("bigscience/bloom-560m");' \
+                'transformers.AutoModel.from_pretrained("openai/whisper-tiny").to("cpu"); transformers.WhisperConfig.from_pretrained("openai/whisper-tiny");transformers.WhisperProcessor.from_pretrained("openai/whisper-tiny");' \
+                'transformers.AutoModel.from_pretrained("google/flan-t5-small").to("cpu"); transformers.AutoTokenizer.from_pretrained("google/flan-t5-small");' > "${IMPORT_SCRIPT_FILENAME}"
```
@huydhn Do you maybe know why caching more models crashes the other builds with `ImportError: /opt/conda/envs/py_3.8/lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/lib/libtorch_python.so)`?

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -406,14 +407,179 @@ def forward(self, x):
+    # TODO: From Config/Model
+    @pytorch_test_common.skip_in_ci(
+        "Skip this test in CI because of memory issue."
+        "SymFloat in OnnxFUnction attribute is not supported yet."
```
https://github.com/pytorch/pytorch/issues/107715

wschin commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -406,14 +407,179 @@ def forward(self, x):
+            onnx.shape_inference.infer_shapes(export_output.model_proto)
+
+    @pytorch_test_common.skip_in_ci(
+        "Skip this test in CI because of memory issue."
```
memory issue? Wasn't it caused by dynamic control flow?

titaiwangms replied:
It's dynamic control-flow. The skip reason is to explain why I didn't put this test under xfail.

wschin replied:
I see, so CI machine doesn't have enough memory to let it fail. Please rephase it to make it clear --- `Not decorated with xfail because CI doesn't have enough memory`.

wschin commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -1012,22 +1055,22 @@ def create_kwargs():
-        def create_kwargs(model_name=model_name):
-            tokenizer = AutoTokenizer.from_pretrained(model_name)
-            return tokenizer("Hello world!", return_tensors="pt")
+        def create_kwargs():
```
In `_test_fake_tensor_mode_exporter`, the exported ONNX model and PyTorch model are run and their outputs are compared. However, I am wondering if that really checks `dynamic_shapes`. If not, probably with the next (or 2nd next), let's modify `create_args`, `create_kwargs`, and `test_fake_tensor_mode_exporter` to guard `dynamic_shapes` supports different batch and sequence sizes.

titaiwangms replied:
There is `assert_dynamic_shapes` inside the `_test_fake_tensor_mode_exporter` should address your concern.

wschin reviewed (approved):
Good first steps to testing important HF models. There are more works required, especially for testing `dynamic_shapes=True` + `FakeTensorMode`, but can be done in another PR.

wschin commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -406,14 +407,179 @@ def forward(self, x):
+            onnx.shape_inference.infer_shapes(export_output.model_proto)
+
+    def test_fake_tensor_mode_huggingface_bigscience_bloom(self):
+        config = transformers.BloomConfig()
```
Explain why not using pre-trained weights.

wschin commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -406,14 +407,179 @@ def forward(self, x):
+        "Please use functorch.experimental.control_flow.cond to explicitly capture the control flow"
+    )
+    def test_fake_tensor_mode_huggingface_mosaicml_mpt_7b(self):
+        model_name = "mosaicml/mpt-7b"
```
Please write the reason that you didn't use `Config` as a comment.

wschin commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -560,21 +560,52 @@ def func(x):
+
+        # Another encoded inputs to test dynamic shapes
+        another_batch, another_seq = 3, 256
+        another_input_ids = torch.randint(0, 8096, (another_batch, another_seq))
```
Merge the input generation code for `[2, 128]` and `[3, 256]` into a function and call it twice.

wschin commented:
Remember to create issues against proper repos (pytorch, hf, onnx, or ort) for the errors/problems encountered during the test.

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -406,14 +407,186 @@ def forward(self, x):
+        model_name = "databricks/dolly-v2-3b"
+        device = "cpu"
+        with torch.onnx.enable_fake_mode() as fake_context:
+            tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)
```
oh, nvm saw skipped in CI.

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -750,14 +782,20 @@ def create_pytorch_only_extra_kwargs():
 
         def create_model() -> nn.Module:
-            return transformers.AutoModel.from_pretrained(model_name)
+            return transformers.AutoModel.from_pretrained(model_name).to(device).eval()
```
this still needs to be cached though right?

titaiwangms replied:
Oops. Added back the tiny-gpt2 cache.

BowenBao reviewed (approved):
LG w/ comment on tests with pretrained model and CI caching.

titaiwangms committed:
Update on "[ONNX] Add huggingface models into CI tests"


1. Add a list of HF models to CI tests. The PR intends to build them from Config, but some of them are not supported with Config. NOTE: Loaded from pre-trained model could potentially hit [uint8/bool conflict](https://github.com/huggingface/transformers/issues/21013) when a newer version of transformers is used.
    - Dolly has torch.fx.Node in OnnxFunction attribute, which is currently not supported.
    - Falcon and MPT has unsupported user coding to Dynamo.
2. Only update GPT2 exporting with real tensor to Config, as FakeMode rises unequal input errors between PyTorch and ORT. The reason is that [non-persistent buffer is not supported](https://github.com/pytorch/pytorch/issues/107211)

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
